### ISSUE TITLE
DeviceManager().enumerate() takes very long

### ISSUE THREAD

**Issue by Sawrz** [abcminiuser/python-elgato-streamdeck#51]
Hi,
This issue happened suddenly after a system upgrade, so it's hard to pinpoint the culprit here. What I found is that `DeviceManager().enumerate()` takes a very long time, over a minute. That was not the case before and makes now streamdeck-ui unusable.
So, I reinstalled all streamdeck python packages and the necessary libraries — still the same issues. Any ideas what causes that behavior?

I'm on Linux Mint 20 and use the 5.8.5-xanmod1 kernel.

Thanks,
Sandro

**Comment by abcminiuser** [comment]
I just wiped and installed a fresh Mint 20 onto my test laptop, as it was running Ubuntu Mate 20.04. Unfortunately I don't see any odd behaviour - after installing the `hidapi-libusb0` dependency the 0.8.2 tag of the library worked with my StreamDeck original just fine.

I'm not sure where to go from here - perhaps a `strace` when running the deck information example script will show what system calls (if any) are stalling out.

Could you try the default Mint 20 kernel as well, to see if it's specific to your custom kernel?

**Comment by Sawrz** [comment]
I did some testing, and it seems not related to the kernel:

I removed the xanmod Kernel, so I run on 5.4.0-42-generic, but it takes still around 1 minute to find the device. 

I'm not familiar with `strace` but I started the example script, got the PID and ran `strace -p <PID>`, according to [this site](https://kracekumar.com/post/168364964555/debugging-python-multiprocessing-program-with/). That is the output:
```
strace: Process 19991 attached
restart_syscall(<... resuming interrupted read ...>) = 1
read(4, "\1", 1)                        = 1
madvise(0x7f20f841e000, 8368128, MADV_DONTNEED) = 0
exit(0)                                 = ?
+++ exited with 0 +++
```

It got stuck here:
```
restart_syscall(<... resuming interrupted read ...>) = 1
```

I can exclude a bad cable or connection by using the API on another machine running Arch - everything works fine ,and I get a nearly instant output. The connection is the same since the Streamdeck is attached to a KVM switch. The only different thing is the physical USB port on the PC. However, there are other devices attached to it that work well.

However, to avoid any noise I replugged the StreamDeck directly to another USB port on the machine where the issue occurs. Still the same result:
```
strace: Process 20237 attached
restart_syscall(<... resuming interrupted read ...>) = 1
read(4, "\1", 1)                        = 1
madvise(0x7f7b78e53000, 8368128, MADV_DONTNEED) = 0
exit(0)                                 = ?
+++ exited with 0 +++
```

It gets stuck at the same place. So, I think we can exclude any physical connection issues.

For the sake of completeness, `pip3 list` outputs: `streamdeck 0.8.2` and my `/etc/udev/rules.d/10-streamdeck.rules`:
```
SUBSYSTEM=="usb", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="0060", MODE:="666", GROUP="plugdev"
SUBSYSTEM=="usb", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="0063", MODE:="666", GROUP="plugdev"
SUBSYSTEM=="usb", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="006c", MODE:="666", GROUP="plugdev"
SUBSYSTEM=="usb", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="006d", MODE:="666", GROUP="plugdev"
```

The additional entries come from the streamdeck-ui configuration.
Does it help you to pinpoint the culprit?

**Comment by abcminiuser** [comment]
Oops, I completely forgot about this, sorry! I'm not sure why, but GitHub only sends me email notifications on *some* issue/PR updates, but not all. Very annoying.

So, that trace looks like the thread is blocking waiting on a `read()` operation to complete, which is eventually timing out. The StreamDeck `enumerate()` function should not actually do any endpoint reads directly - it just calls into the `hidapi` library's enumeration function with a VID and PID device filter to get a linked list of matching devices. Those list entries may trigger read calls however, as it needs to obtain the device and string descriptors in order to populate the device details.

Are you definitely using the 0.8.2 or later library version? Do you have any other software running that might be using the StreamDeck device?

**Comment by Sawrz** [comment]
I tried to install it to see the version and I got:
``` bash
libhidapi-libusb0 is already the newest version (0.9.0+dfsg-1)
```

I assume that's the `hidapi-libusb0` dependency. 

No I don't use any software which could use the StreamDeck device. The weird part is, it suddenly stopped working. I thought it might broke due an update.

**Comment by Sawrz** [comment]
Hi,
I installed streamdeck this time on my Arch installation. Initially, it worked flawlessly. However, after the first usage I run into the same errors. This time I installed it via `yay`.
I don't want to push you but just keeping you up to date. =)

**Comment by Sawrz** [comment]
I got a second Streamdeck, and the problem persists. However, on one of my systems, the deck works, but the issue is still the same on the other (where I attached my old one). 

I removed the .streamdeck_ui.json to see if anything changes, but the problem is still there. Are there any other configuration/temporary files worth deleting? It seems something goes wrong on that end. 

I reinstalled both `python-elgato-streamdeck` and `streamdeck-ui` via `yay`.
